Feature/ratelimit middleware

bluwumeaway opened the pull request:
uses `/x/time/rate` which was already a dependency for the implementation of the ratelimit. there are separate rate limiters for both the messages inside of the websocket vs "normal" api endpoints. From my testing the current limits seem to work/not cause issues with the site in development.

bluwumeaway committed:
Test for websocket ratelimits.

bluwumeaway committed:
Test for websocket ratelimits.

bluwumeaway committed:
parse out port

bluwumeaway committed:
ope fix formatting

bluwumeaway committed:
Merge branch 'streamplace:next' into feature/ratelimit

bluwumeaway committed:
move ratelimit to middleware

bluwumeaway committed:
Merge branch 'next' of github.com:bluwumeaway/streamplace into feature/ratelimit-middleware

bluwumeaway committed:
bump ratelimit size, remove extra net

bluwumeaway committed:
Merge branch 'feature/ratelimit-middleware' of github.com:bluwumeaway/streamplace into feature/ratelimit-middleware

iameli commented on pkg/api/api.go:
```diff
@@ -669,6 +682,28 @@ func (a *StreamplaceAPI) HandleLivestream(ctx context.Context) httprouter.Handle
+func (a *StreamplaceAPI) RateLimitMiddleware(ctx context.Context) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			// XXX: check if x-forwarded-for
```
not a concern for streamplace in production, we host our own HTTPS, but not impossible someone else could configure a proxy for sure

iameli commented on pkg/api/websocket.go:
```diff
@@ -203,6 +210,27 @@ func (a *StreamplaceAPI) HandleWebsocket(ctx context.Context) httprouter.Handle
+					return
+				}
+			}
+
```
Not worried about this case - this is user --> streamplace messages, and zero of those exist right now - but we should have an IP-specific check for "number of websockets currently open" to handle that one person that's spamming us

iameli commented on pkg/api/api.go:
```diff
@@ -669,6 +682,28 @@ func (a *StreamplaceAPI) HandleLivestream(ctx context.Context) httprouter.Handle
+			limiter := a.getLimiter(ip)
+
+			if !limiter.Allow() {
+				log.Warn(ctx, "rate limit exceeded", "ip", ip, "path", req.URL.Path)
```
note to self: add to grafana

iameli reviewed (changes requested):
Really great start! The one missing piece is "open websockets per IP address" which I don't think will be handled by the "requests per second" handler; we'd need something that counts up when a websocket upgrades and counts down when the websocket closes

bluwumeaway committed:
remove websocket message rate limts add maxConns per IP

bluwumeaway commented:
I removed the to server ratelimiting, and added a conn tracker, but I think it should be configured via the `config` package instead of hardcoded to 5. but i dont know what it would be called in that and it is kind of confusing

iameli reviewed (approved):
LGTM, will merge when CI passes. TY!
